src/backend/gpopt/CGPOptimizer.cpp

```
@@ -61,11 +61,20 @@ CGPOptimizer::GPOPTOptimizedPlan(
 	}
 	GPOS_CATCH_EX(ex)
 	{
-		// clone the error message before context free.
-		CHAR *serialized_error_msg =
-			gpopt_context.CloneErrorMsg(MessageContext);
-		// clean up context
-		gpopt_context.Free(gpopt_context.epinQuery, gpopt_context.epinPlStmt);
+		CHAR *serialized_error_msg = NULL;
+		GPOS_TRY
+		{
+			// clone the error message before context free.
+			serialized_error_msg =
+				gpopt_context.CloneErrorMsg(MessageContext);
+			// clean up context
+			gpopt_context.Free(gpopt_context.epinQuery, gpopt_context.epinPlStmt);
+		}
+		GPOS_CATCH_EX(ex)
+		{
+			PG_RE_THROW();
+		}
+		GPOS_CATCH_END;
```

Does catching and rethrowing here cause us to swallow the original exception?

It looks like `serialized_error_msg` is only used inside the nested if block below.

```cpp
if (optimizer_trace_fallback)
{
	if (errstart(INFO, TEXTDOMAIN))
	{
		// ...
```

Maybe we should move the message copy and all references to `serialized_error_msg` into that block?

---

Yes, you are right. Rethrowing here will swallow the original exception.
I've also considered your suggestion, but things here are a little complex, because when `CloneErrorMsg` raised an ERROR, there should indicates a memory allocation issue, and i think throw this memory exception is more important than throwing ORCA's exception for the end user, as it points to a more critical underlying problem.

---

If memory allocation is the issue, then does the original exception not manifest that issue, too?

IIUC you're suggesting that we use `CloneErrorMsg()` as a test for memory allocation issues.

---

Agreed that alloc memory exception is more important.

But could the original exception be a alloc memory exception?  If yes, then I think the proposed code change would trigger the same exception twice.  But even if it can't, why should we use `gpopt_context.CloneErrorMsg` as a test in case it might be just enough to tip the scales?

I was thinking something like following (disclaimer I have not tested it):
```diff
diff --git a/src/backend/gpopt/CGPOptimizer.cpp b/src/backend/gpopt/CGPOptimizer.cpp
index e1d4ae40591..b1a442a8022 100644
--- a/src/backend/gpopt/CGPOptimizer.cpp
+++ b/src/backend/gpopt/CGPOptimizer.cpp
@@ -61,12 +61,6 @@ CGPOptimizer::GPOPTOptimizedPlan(
        }
        GPOS_CATCH_EX(ex)
        {
-               // clone the error message before context free.
-               CHAR *serialized_error_msg =
-                       gpopt_context.CloneErrorMsg(MessageContext);
-               // clean up context
-               gpopt_context.Free(gpopt_context.epinQuery, gpopt_context.epinPlStmt);
-
                // Special handler for a few common user-facing errors. In particular,
                // we want to use the correct error code for these, in case an application
                // tries to do something smart with them.
@@ -97,23 +91,34 @@ CGPOptimizer::GPOPTOptimizedPlan(
                {
                        if (errstart(INFO, TEXTDOMAIN))
                        {
-                               errcode(ERRCODE_FEATURE_NOT_SUPPORTED);
-                               errmsg(
-                                       "GPORCA failed to produce a plan, falling back to Postgres-based planner");
-                               if (serialized_error_msg)
+                               GPOS_TRY
                                {
-                                       errdetail("%s", serialized_error_msg);
+                                       // clone the error message before context free.
+                                       CHAR *serialized_error_msg =
+                                               gpopt_context.CloneErrorMsg(MessageContext);
+                                       // clean up context
+                                       gpopt_context.Free(gpopt_context.epinQuery,
+                                                                          gpopt_context.epinPlStmt);
+
+                                       errcode(ERRCODE_FEATURE_NOT_SUPPORTED);
+                                       errmsg(
+                                               "GPORCA failed to produce a plan, falling back to Postgres-based planner");
+                                       if (serialized_error_msg)
+                                       {
+                                               errdetail("%s", serialized_error_msg);
+                                               pfree(serialized_error_msg);
+                                       }
+                                       errfinish(ex.Filename(), ex.Line(), nullptr);
                                }
-                               errfinish(ex.Filename(), ex.Line(), nullptr);
+                               GPOS_CATCH_EX(ex)
+                               {
+                                       PG_RE_THROW();
+                               }
+                               GPOS_CATCH_END;
                        }
                }
 
                *had_unexpected_failure = gpopt_context.m_is_unexpected_failure;
-
-               if (serialized_error_msg)
-               {
-                       pfree(serialized_error_msg);
-               }
        }
```

I think in that case, if the original exception is already a memory alloc exception then we'd propagate that up right away.  However, if it's a fallback that leads to a memory alloc exception, then we'd raise only the latter exception.